contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,768 @@
+    
+    /**
+     * Runs all the necessary approval functions required for a list of ERC20 tokens.
+     *
+     * @param _tokens    Addresses of the tokens which need approval
+     */
+    function approveTokens(IERC20[] calldata _tokens) external {
+        for (uint256 i = 0; i < _tokens.length; i++)
+            approveToken(_tokens[i]);
+    }
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a SetToken. This function need to be called only once before the first time
+     * this smart contract is used on any particular SetToken.
+     *
+     * @param _setToken    Address of the SetToken being initialized
+     */
+    function approveSetToken(ISetToken _setToken) external {
+        address[] memory components = _setToken.getComponents();
+        for (uint256 i = 0; i < components.length; i++)
+            approveToken(IERC20(components[i]));
+    }
+
+    /**
+     * Issues SetTokens for an exact amount of input ERC20 tokens.
+     * The ERC20 token must be approved by the sender to this contract. 
+     *
+     * @param _setToken         Address of the SetToken being issued
+     * @param _inputToken       Address of input token
+     * @param _amountInput      Amount of the input token / ether to spend
+     * @param _minSetReceive    Minimum amount of SetTokens to receive. Prevents unnecessary slippage.
+     */
+    function issueSetForExactToken(
+        ISetToken _setToken,
+        IERC20 _inputToken,
+        uint256 _amountInput,
+        uint256 _minSetReceive
+    )   
+        isSetToken(_setToken)
+        external
+        nonReentrant
+    {   
+        require(_amountInput > 0, "ExchangeIssuance: INVALID INPUTS");
+        
+        _inputToken.safeTransferFrom(msg.sender, address(this), _amountInput);
+        
+        if(address(_inputToken) != WETH) 
+           _swapTokenForWETH(_inputToken, _amountInput);
+
+        uint256 setTokenAmount = _issueSetForExactWETH(_setToken, _minSetReceive);
+        
+        emit ExchangeIssue(msg.sender, _setToken, address(_inputToken), _amountInput, setTokenAmount);
+    }
+    
+    /**
+     * Issues SetTokens for an exact amount of input ether.
+     * 
+     * @param _setToken         Address of the SetToken to be issued
+     * @param _minSetReceive    Minimum amount of SetTokens to receive. Prevents unnecessary slippage.
+     */
+    function issueSetForExactETH(
+        ISetToken _setToken,
+        uint256 _minSetReceive
+    )
+        isSetToken(_setToken)
+        external
+        payable
+        nonReentrant
+    {
+        require(msg.value > 0, "ExchangeIssuance: INVALID INPUTS");
+        
+        IWETH(WETH).deposit{value: msg.value}();
+        
+        uint256 setTokenAmount = _issueSetForExactWETH(_setToken, _minSetReceive);
+        
+        emit ExchangeIssue(msg.sender, _setToken, address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE), msg.value, setTokenAmount);   
```

Comment:
All constants should be declared at the top of the file for clarity. We want to avoid magic values!

This constant should be named as `ETH_ADDRESS`